chain/client/src/stateless_validation/endorsement_tracker.rs

```diff
@@ -0,0 +1,141 @@
+            tracing::error!(target: "stateless_validation", ?endorsement, "Invalid chunk endorsement.");
+            return Err(Error::InvalidChunkEndorsement);
+        }
+
+        // If we are the current block producer, we store the chunk endorsement for each chunk which
+        // would later be used during block production to check whether to include the chunk or not.
+        // TODO(stateless_validation): It's possible for a malicious validator to send endorsements
+        // for 100 unique chunks thus pushing out current valid endorsements from our cache.
+        // Maybe add check to ensure we don't accept endorsements from chunks already included in some block?
+        // Maybe add check to ensure we don't accept endorsements from chunks that have too old height_created?
+        tracing::debug!(target: "stateless_validation", ?endorsement, "Received and saved chunk endorsement.");
+        self.chunk_endorsements.get_or_insert(chunk_hash.clone(), || HashMap::new());
+        let chunk_endorsements = self.chunk_endorsements.get_mut(chunk_hash).unwrap();
+        chunk_endorsements.insert(account_id.clone(), endorsement);
+
+        Ok(())
+    }
+
+    /// Called by block producer.
+    /// Returns Some(signatures) if node has enough signed stake for the chunk represented by chunk_header.
+    /// Signatures have the same order as ordered_chunk_validators, thus ready to be included in a block as is.
+    /// Returns None if chunk doesn't have enough stake.
+    /// For older protocol version, we return an empty array of chunk endorsements.
+    pub fn get_chunk_endorsement_signatures(
+        &self,
+        chunk_header: &ShardChunkHeader,
+    ) -> Result<Option<ChunkEndorsementSignatures>, Error> {
+        let epoch_id =
+            self.epoch_manager.get_epoch_id_from_prev_block(chunk_header.prev_block_hash())?;
+        let protocol_version = self.epoch_manager.get_epoch_protocol_version(&epoch_id)?;
+        if !checked_feature!("stable", ChunkValidation, protocol_version) {
+            // Return an empty array of chunk endorsements for older protocol versions.
+            return Ok(Some(vec![]));
+        }
+
+        // Get the chunk_endorsements for the chunk from our cache.
+        // Note that these chunk endorsements are already validated as part of process_chunk_endorsement.
+        // We can safely rely on the the following details
+        //    1. The chunk endorsements are from valid chunk_validator for this chunk.
+        //    2. The chunk endorsements signatures are valid.
+        let Some(chunk_endorsements) = self.chunk_endorsements.peek(&chunk_header.chunk_hash())
+        else {
+            // Early return if no chunk_enforsements found in our cache.
+            return Ok(None);
+        };
+
+        let epoch_id =
+            self.epoch_manager.get_epoch_id_from_prev_block(chunk_header.prev_block_hash())?;
```

Comment:
you already have epoch_id variable

Reply:
Good catch! Might have been introduced while merging from master in some previous PR